docs/json-map.md

```diff
@@ -0,0 +1,72 @@
+# Overview
+JSON Map is a JSON-based format which describes part of the structure of another JSON document. Maps inform clients of the sizes of parts of the JSON and to allow constant time construction of subsets of the document.
+
+For example, suppose you have a 20GB JSON file which contains a huge array with one million elements along with some surrounding document structure. You can construct a map with a 0.1% target size (20 MB). The map will identify the path to the huge array, the number of elements it has, where each nth array element begins, and where the array itself starts and ends. 
+
+To write a copy of the file with only the first 1,000 array elements, copy the file content up to the array start, copy the range from element 0 to element 999, omit the last comma, and copy the range from the array end to the end of the file.
+
+
+# JSON Map Structure
+A JSON Map consists of a hierarchy of **nodes** which mirror the hierarchical structure of the underlying JSON document. A map is constructed with a specific **target size ratio** (ex: 1%) which determines a size threshold for including items in the map.
+
+Each **node** contains:
+* A "start" integer, which is the absolute byte position in the file of the start of the object or array (the index of the '[' or '{')
+* An "end" integer, which is the absolute byte position in the file of the end of the object or array (the index of the ']' or '}')
```

Comment:
>end [](start = 6, length = 3)

Just checking -- "end" really points to the closing brace, rather than one past it? In iterators, the "end" iterator conventionally points one past, so that "end" - "start" == length. Not suggesting you change it; just making sure this is really what you meant.

Reply:
Yes, that's the intent. It's easier to get the position of a token ('{') from the parser/reader than the position after it. The position after it also may or may not be whitespace, depending on whether the JSON is indented, causing other weirdness when concatenating subsets of JSON from these offsets.